src/Dom/ResourceHintManager.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+namespace AmpProject\Dom;
+
+use AmpProject\Attribute;
+use AmpProject\RequestDestination;
+use AmpProject\Tag;
+use DOMNode;
+
+/**
+ * Resource hint manager class that is used to manage the <link> tags within a document's <head> to give the browser
+ * hints about how to prioritize resources.
+ *
+ * @package ampproject/amp-toolbox
+ */
+final class ResourceHintManager
+{
+
+    /**
+     * Document to manage the resource hints for.
+     *
+     * @var Document
+     */
+    private $document;
+
+    /**
+     * Reference node to attach the resource hint to.
+     *
+     * @var DOMNode|null
+     */
+    private $referenceNode;
+
+    /**
+     * ResourceHintManager constructor.
+     *
+     * @param Document $document
+     */
+    public function __construct(Document $document)
+    {
+        $this->document = $document;
+    }
+
+    /**
+     * Add a preconnect resource hint.
+     *
+     * @param string $href        URL to link to.
+     * @param bool   $crossorigin Whether the link is crossorigin.
+     */
+    public function addPreconnect($href, $crossorigin = true)
+    {
+        $this->addLinkTag(
+            Attribute::REL_PRECONNECT,
+            $href,
+            $crossorigin ? [ Attribute::CROSSORIGIN => null ] : []
+        );
+
+        // Use dns-prefetch as fallback for browser that don't support preconnect.
+        // See https://web.dev/preconnect-and-dns-prefetch/#resolve-domain-name-early-with-reldns-prefetch
+        $this->addLinkTag(Attribute::REL_DNS_PREFETCH, $href);
+    }
+
+    /**
+     * Add a preconnect resource hint.
+     *
+     * @param string      $href  URL to link to.
+     * @param string      $type  Optional. Type of the resource. Defaults to type 'image'.
+     * @param string|null $media Optional. Media query to add to the preload. Defaults to none.
+     */
+    public function addPreload($href, $type = RequestDestination::IMAGE, $media = null)
```

Comment:
Technically preload links are not a [resource hints](https://www.w3.org/TR/resource-hints/). It's a [separate thing](https://www.w3.org/TR/preload/). See https://core.trac.wordpress.org/ticket/42438#comment:3

This this reason the note about `private` below, what about changing `ResourceHintManager` to be rather `LinkManager`. Then this could be used for adding any `link`.